internal/iamauth/util.go

```diff
@@ -0,0 +1,167 @@
+package iamauth
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/hashicorp/consul/internal/iamauth/responses"
+	"github.com/hashicorp/go-hclog"
+)
+
+type LoginInput struct {
+	Creds            *credentials.Credentials
+	IncludeIAMEntity bool
+	STSEndpoint      string
+	STSRegion        string
+
+	Logger hclog.Logger
+
+	ServerIDHeaderValue string
+	// Customizable header names
+	ServerIDHeaderName     string
+	GetEntityMethodHeader  string
+	GetEntityURLHeader     string
+	GetEntityHeadersHeader string
+	GetEntityBodyHeader    string
+}
+
+// GenerateLoginData populates the necessary data to send for the bearer token.
+// https://github.com/hashicorp/go-secure-stdlib/blob/main/awsutil/generate_credentials.go#L232-L301
+func GenerateLoginData(in *LoginInput) (map[string]interface{}, error) {
+
+	// Use the credentials we've found to construct an STS session
+	// TODO: Do we need this? Or can just assume the region from the credentials?
+	// region, err := GetRegion(configuredRegion)
```

Comment:
TODO: Need to dive into this. Relates to my other comment about a custom credential chain vs using the AWS SDK's standard credential chain.